upgrade/upgrades: TestSessionBasedLeaseUpgrade failed

upgrade/upgrades.TestSessionBasedLeaseUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14553873?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14553873?buildTab=artifacts#/) on master @ [7674256ec5c1447c7c5fe8f7538d319466474811](https://github.com/cockroachdb/cockroach/commits/7674256ec5c1447c7c5fe8f7538d319466474811):


```
=== RUN   TestSessionBasedLeaseUpgrade
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/19fe8462015815b7be025b680724a827/logTestSessionBasedLeaseUpgrade555063793
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:144: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    v24_1_session_based_lease_test.go:104: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/v24_1_session_based_lease_test.go:104
        	Error:      	Received unexpected error:
        	            	pq: transaction committed but schema change aborted with error: (42703): count-leases: column "desc_id" does not exist
        	Test:       	TestSessionBasedLeaseUpgrade
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/19fe8462015815b7be025b680724a827/logTestSessionBasedLeaseUpgrade555063793
--- FAIL: TestSessionBasedLeaseUpgrade (2.48s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/release-eng
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSessionBasedLeaseUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-37039

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -1417,7 +1417,7 @@ func (ex *connExecutor) checkDescriptorTwoVersionInvariant(ctx context.Context) 
 	if knobs := ex.server.cfg.SchemaChangerTestingKnobs; knobs != nil {
 		inRetryBackoff = knobs.TwoVersionLeaseViolation
 	}
-	regionCache, err := regions.NewCachedDatabaseRegions(ctx, ex.server.cfg.DB, ex.server.cfg.LeaseManager)
+	regionCache, err := regions.NewCachedDatabaseRegionsAt(ctx, ex.server.cfg.DB, ex.server.cfg.LeaseManager, ex.state.mu.txn.ProvisionalCommitTimestamp())
 	if err != nil {
 		return err
 	}

--- a/pkg/sql/regions/BUILD.bazel
+++ b/pkg/sql/regions/BUILD.bazel
@@ -17,6 +17,7 @@ go_library(
         "//pkg/sql/catalog",
         "//pkg/sql/catalog/descs",
         "//pkg/sql/catalog/lease",
+        "//pkg/util/hlc",
         "@com_github_cockroachdb_errors//:errors",
     ],
 )

--- a/pkg/sql/regions/cached_db_regions.go
+++ b/pkg/sql/regions/cached_db_regions.go
@@ -18,6 +18,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/lease"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 )
 
 // CachedDatabaseRegions tracks the regions that are part of
@@ -27,14 +28,14 @@ type CachedDatabaseRegions struct {
 	dbRegionEnumDesc catalog.TypeDescriptor
 }
 
-// NewCachedDatabaseRegions creates a new region cache by leasing
-// the underlying descriptors for the system database and region enum.
-func NewCachedDatabaseRegions(
-	ctx context.Context, db *kv.DB, lm *lease.Manager,
+// NewCachedDatabaseRegionsAt creates a new region cache by leasing
+// the underlying descriptors for the system database and region enum,
+// with descriptors fetched at a specific timestamp.
+func NewCachedDatabaseRegionsAt(
+	ctx context.Context, _ *kv.DB, lm *lease.Manager, at hlc.Timestamp,
 ) (cdr *CachedDatabaseRegions, err error) {
 	cdr = &CachedDatabaseRegions{}
-	now := db.Clock().Now()
-	desc, err := lm.Acquire(ctx, now, keys.SystemDatabaseID)
+	desc, err := lm.Acquire(ctx, at, keys.SystemDatabaseID)
 	if err != nil {
 		return nil, err
 	}
@@ -48,13 +49,21 @@ func NewCachedDatabaseRegions(
 		return nil, err
 	}
 
-	desc, err = lm.Acquire(ctx, now, enumID)
+	desc, err = lm.Acquire(ctx, at, enumID)
 	if err != nil {
 		return nil, err
 	}
 	defer desc.Release(ctx)
 	cdr.dbRegionEnumDesc = desc.Underlying().(catalog.EnumTypeDescriptor)
 	return cdr, nil
+}
+
+// NewCachedDatabaseRegions creates a new region cache by leasing
+// the underlying descriptors for the system database and region enum.
+func NewCachedDatabaseRegions(
+	ctx context.Context, db *kv.DB, lm *lease.Manager,
+) (*CachedDatabaseRegions, error) {
+	return NewCachedDatabaseRegionsAt(ctx, db, lm, db.Clock().Now())
 }
 
 // IsMultiRegion returns if the system database is multi-region.

